Implement new await synchronous interface

ellnix opened the pull request:
# Pull Request

## Related issue
Fixes #288

While refactoring, inadvertently fixes:
- Fixes #499
- Fixes #500

## PR checklist
Please check if your PR fulfills the following requirements:
- [X] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [X] Have you read the contributing guidelines?
- [X] Have you made sure that the title is accurate and descriptive of the changes?

## Changes Checklist
- [X] Add soft deprecation messages
- [X] Remove bang methods from specs
- [X] Add Task "model"
- [X] Ensure all tests pass with edits where necessary
- [X] Add tests to make sure that all bang methods actually provide deprecation warnings 
- [X] Return Task model on all methods that return tasks
  - [X] Return Task model from all settings setters and getters
- [ ] Update the documentation with new syntax (if applicable)

ellnix commented:
@brunoocasali this is pretty much complete, please feel free to review when you have (a lot of) time.

ellnix commented:
Rebased and force pushed.

brunoocasali reviewed (commented):
Thanks a lot for your hard work here @ellnix amazing!

I am asking for some changes but it is nothing to be worried about!

brunoocasali reviewed (changes requested):
Amazing work @ellnix I think we could make some small modifications, but I was impressed, very good work!

ellnix commented:
@brunoocasali sorry for taking so long to get back to this. I addressed your feedback and rebased it to main (for the changes since you last reviewed it, start at commit [b141ed4](https://github.com/meilisearch/meilisearch-ruby/pull/498/commits/b141ed48198449090f93ede4d7281ae65b01960b)).

brunoocasali commented:
let's see if the code coverage will still be broken now!

Also, can you squash the commits @ellnix ?

ellnix commented:
> Also, can you squash the commits @ellnix ?

Squash them into a single commit or just fewer commits? How many commits would you consider ideal? The history might be a little bit hard to follow if only a few commits had +1800 -1200 changes.

brunoocasali commented on lib/meilisearch/utils.rb:
```diff
@@ -4,70 +4,81 @@ module MeiliSearch
 
-      message = <<~MSG
-        Attributes will be expected to be snake_case in future versions of Meilisearch Ruby.
+        logger.warn(message)
```
> I took the liberty to unify the warning logging with a Rails-like logger. 

Good for me!

`MeiliSearch.logger` is better indeed!

brunoocasali commented:
> Squash them into a single commit or just fewer commits? How many commits would you consider ideal? The history might be a little bit hard to follow if only a few commits had +1800 -1200 changes.

A few commits would be great. I don't have a number for that, it is more a feeling than anything. If you squash them based on the group like, you did a lot of the same but in multiple files, so in my mind you would have a single commit of that "theme".

ellnix commented:
@brunoocasali should be ready for merge, unless you want me to squash it down further.

brunoocasali commented on spec/meilisearch/models/task_spec.rb:
```diff
@@ -0,0 +1,485 @@
+      it { is_expected.not_to be_enqueued }
+
+      it 'does not refresh the task' do
+        allow(subject).to receive(:refresh)
+        subject.enqueued?
+        expect(subject).not_to have_received(:refresh)
```
When I mention the xUnit I mean this:

```suggestion
        allow(subject).to receive(:refresh)
        
        subject.enqueued?
        
        expect(subject).not_to have_received(:refresh)
```

brunoocasali reviewed (approved):
Looks good to me, let get this merged @ellnix

Thanks a lot for all the hard work in this repo @ellnix ❤️

ellnix commented:
I'm not sure this will work, but addressing the conflict with a commit might dismiss the approval review.

bors merge

ellnix commented:
Give me a second with this, apparently the git pull on my fork didn't go through properly and now everything is messed up.

ellnix commented:
Well I fixed the git history and force pushed. Sorry about the issues but I don't think there was a way for me to resolve the main conflicts without automatically dismissing the approval review (even when I tried a merge rather than a rebase).

I looked it up, it seems to be a setting (https://github.com/isaacs/github/issues/783#issuecomment-320211882) but it does not look like it would be a good idea, since it would allow someone with approval to push arbitrary code and then choose to merge.

brunoocasali reviewed (approved):
LGTM!

ellnix commented:
bors merge
